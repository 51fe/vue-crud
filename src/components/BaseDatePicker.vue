<template>
  <el-date-picker
    :value="value"
    type="date"
    value-format="yyyy-MM-dd HH:mm:ss"
    :placeholder="placeholder"
    clearable
    v-bind="$attrs"
    class="base-date-picker"
    @input="$emit('input', $event)"
  />
</template>

<script>
export default {
  name: 'BaseDatePicker',
  props: {
    value: {
      type: [String, Date],
      default: ''
    },
    placeholder: {
      type: String,
      default: '请选择'
    }
  }
}
</script>

<style lang="less">
  .base-date-picker {
    &.el-date-editor {
      &.el-input, .el-input__wrapper {
        width: 100%;
      }
    }
  }
  </style>
