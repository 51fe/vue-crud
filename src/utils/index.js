/**
 * Parse the time to string
 * @param {(Date|string|number)} time
 * @param {string} format
 * @returns {string | null}
 */
export function parseDateTime(time, format ='{y}-{m}-{d} {h}:{i}:{s}') {
  if (arguments.length === 0 || !time) {
    return null
  }
  let date
  if (typeof time === 'object') {
    date = time
  } else {
    if ((typeof time === 'string')) {
      if ((/^[0-9]+$/.test(time))) {
        // support "1548221490638"
        time = parseInt(time)
      } else {
        // support safari
        // https://stackoverflow.com/questions/4310953/invalid-date-in-safari
        time = time.replace(new RegExp(/-/gm), '/')
      }
    }

    if ((typeof time === 'number') && (time.toString().length === 10)) {
      time = time * 1000
    }
    date = new Date(time)
  }
  const formatObj = {
    y: date.getFullYear(),
    m: date.getMonth() + 1,
    d: date.getDate(),
    h: date.getHours(),
    i: date.getMinutes(),
    s: date.getSeconds(),
    a: date.getDay()
  }
  const time_str = format.replace(/{([ymdhisa])+}/g, (result, key) => {
    const value = formatObj[key]
    // Note: getDay() returns 0 on Sunday
    if (key === 'a') { return ['一', '二', '三', '四', '五', '六', '日'][value ] }
    return value.toString().padStart(2, '0')
  })
  return time_str
}

/**
 * 根据值获取标签
 * @param list
 * @param value
 * @returns {string|*}
 */
export function getLabel(list, value) {
  const found = list?.find((item) => item.value == value)
  if (found) return found.label
  return ''
}

/**
 * 根据区码查找区名
 * @param list
 * @param value
 * @returns {string|*}
 */
export function getAreaNameByCode(list, value) {
  if(Array.isArray(list)) {
    for (const item of list) {
      if (item.value === value) {
        return item.label
      } else if (item.children?.length > 0) {
        return getAreaNameByCode(item.children, value)
      }
    }
  }
  return ''
}

/**
 * 简单深度拷贝
 * @param data
 * @returns {any}
 */
export function deepClone(data) {
  return JSON.parse(JSON.stringify(data))
}
