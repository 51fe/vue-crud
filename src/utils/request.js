import axios from 'axios'
import { Message } from 'element-ui'

// 创建axios实例
const service = axios.create({
  baseURL: process.env.VUE_APP_BASE_API
})

// request拦截器
service.interceptors.request.use(
  config => {
    const method = config.method.toLocaleLowerCase()
    if(['post', 'put'].includes(method)) {
      config.headers['Content-Type'] = 'application/json'
    }
    return config
  },
  error => {
    // Do something with request error
    console.log(error) // for debug
    Promise.reject(error)
  }
)

// response 拦截器
service.interceptors.response.use(
  response => {
    const { data, status, config, headers, statusText } = response
    if (status < 200 || status > 300) {
      Message({
        message: statusText,
        type: 'error',
        duration: 5 * 1000
      })
    } else {
      const action = getAction(config, status)
      if(action) {
        Message({
          message: action + '成功',
          type: 'success'
        })
      }
      const count = headers['x-total-count']
      if(count) {
        return {
          list: data,
          total: parseInt(count)
        }
      }
      return data
    }
  },
  error => {
    Message({
      message: error.message ?? '未知错误',
      type: 'error',
      duration: 5 * 1000
    })
    return Promise.reject(error.code)
  }
)

function getAction(config, status) {
  let action = ''
  if(status === 201) {
    action = '新增'
  } else {
    const method = config?.method?.toLocaleLowerCase() ?? ''
    if (method === 'put') {
      action = '修改'
    } else if (method === 'delete') {
      action = '删除'
    }
  }
  return action
}

export default service
