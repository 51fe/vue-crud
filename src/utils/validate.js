/**
 * 检验邮编
 * @param rule
 * @param value
 * @param callback
 */
export function zipValidator(rule, value, callback) {
  const reg = /(^\d{6}$)/
  if (value) {
    if (!reg.test(value)) {
      callback(new Error('请输入正确的邮编'))
    } else {
      callback()
    }
  } else {
    callback()
  }
}
