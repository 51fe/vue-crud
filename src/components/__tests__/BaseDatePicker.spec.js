import { fireEvent, render, screen } from '@testing-library/vue'
import BaseDatePicker from '../BaseDatePicker.vue'
import { parseDateTime } from '@/utils'

test('renders correctly by default', () => {
  const { html } = render(BaseDatePicker)
  screen.getByPlaceholderText('请选择')
  expect(html()).toMatchSnapshot()
})

test('changes props correctly', () => {
  const value = '2022-08-02'
  const placeholder = '开始日期'
  render(BaseDatePicker, {
    props: {
      value,
      placeholder
    }
  })
  screen.getByPlaceholderText(placeholder)
  screen.getByDisplayValue(value)
})

test('emits input event correctly', async () => {
  const day = 15
  const { emitted } = render(BaseDatePicker)
  // opens the dropdown menu by focus the input
  await fireEvent.focus(screen.getByRole('textbox'))
  const today = new Date()
  const cell = screen.getByRole('cell', { name: day})
  // emits the event when click the date button
  await fireEvent.click(cell)
  const selected = today.setDate(day)
  expect(emitted().input[0][0]).toBe(parseDateTime(selected, '{y}-{m}-{d} 00:00:00'))
})

