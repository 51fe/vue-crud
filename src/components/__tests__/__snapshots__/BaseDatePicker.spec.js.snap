// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly by default 1`] = `
<div class="el-date-editor base-date-picker el-input el-input--prefix el-input--suffix el-date-editor--date">
  <!----><input type="text" autocomplete="off" name="" placeholder="请选择" class="el-input__inner"><span class="el-input__prefix"><i class="el-input__icon el-icon-date"></i><!----></span><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-input__icon"></i><!----><!----><!----><!----></span>
  <!----></span>
  <!---->
  <!---->
</div>
`;
